Fix to issue #4606 : Links leading outside Studio need to have a pop out icon

shivam-daksh opened the pull request:
<!-- Please remove any unused sections.

Note that anything written between these symbols will not appear in the actual, published PR. They serve as instructions for filling out this template. You may want to use the 'preview' tab above this textbox to verify formatting before submitting.
-->

## Summary
### Description of the change(s) you made
<!-- Briefly summarize your changes in 1-2 sentences here. -->
This merge request addresses the issue where external links in Studio did not have a visual indicator to distinguish them from internal links. The affected links also did not open in a new tab by default. To resolve this, the **KExternalLink** component has been replaced with the **ActionLink** component, and a CSS class **inline-icon** has been added to provide the necessary visual indication.

### Manual verification steps performed
1. Step 1 : Replaced instances of `KExternalLink` with `ActionLink` to ensure consistent behavior and appearance across the application.
2. Step 2 : Added the `inline-icon` CSS class to the ActionLink component to display a pop-out icon for adjusting the position.
3. ...

### Screenshots (if applicable)
<!-- If not applicable, please delete this section -->

https://github.com/user-attachments/assets/4db79b12-d93c-40a7-a540-77411d438270



## Affected Areas

- **In Settings -> page -> Account -> Index.vue : ** Updated the "API documentation" link to use the `ActionLink` component with the `inline-icon` class.
- **In Settings -> page -> UsingStudio -> index.vue :** Updated the "User Guide", "bestPractice6", "issueLink1", "issuesPageLink" link to use the `ActionLink` component replacing **KExternalLink**  component and  adding the `inline-icon` class rep.

## Reviewer guidance
### How can a reviewer test these changes?

1. Go to Studio then Settings.
2. Check all the links leading outside the studio have pop-out icon (open in new tab) in all three tabs.

### Contributor's Checklist
<!-- After saving the PR, come through to tick off completed checklist items. Delete any sections that are not applicable to your PR -->

PR process:

- [ ] If this is an important user-facing change, PR or related issue the `CHANGELOG` label been added to this PR. Note: items with this label will be added to the [CHANGELOG](https://github.com/learningequality/studio/blob/master/CHANGELOG.md) at a later time
- [ ] If this includes an internal dependency change, a link to the diff is provided
- [ ] The `docs` label has been added if this introduces a change that needs to be updated in the [user docs](https://kolibri-studio.readthedocs.io/en/latest/index.html)?
- [ ] If any Python requirements have changed, the updated `requirements.txt` files also included in this PR
- [ ] Opportunities for using Google Analytics here are noted
- [ ] Migrations are [safe for a large db](https://www.braintreepayments.com/blog/safe-operations-for-high-volume-postgresql/)

Studio-specifc:

- [ ] All user-facing strings are translated properly
- [ ] The `notranslate` class been added to elements that shouldn't be translated by Google Chrome's automatic translation feature (e.g. icons, user-generated text)
- [ ] All UI components are LTR and RTL compliant
- [ ] Views are organized into `pages`, `components`, and `layouts` directories [as described in the docs](https://github.com/learningequality/studio/blob/vue-refactor/docs/architecture.md#where-does-the-frontend-code-live)
- [ ] Users' storage used is recalculated properly on any changes to main tree files
- [ ] If there new ways this uses user data that needs to be factored into our [Privacy Policy](https://github.com/learningequality/studio/tree/master/contentcuration/contentcuration/templates/policies/text), it has been noted.


Testing:

- [ ] Code is clean and well-commented
- [ ] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [ ] Any new interactions have been added to the [QA Sheet](https://docs.google.com/spreadsheets/d/1HF4Gy6rb_BLbZoNkZEWZonKFBqPyVEiQq4Ve6XgIYmQ/edit#gid=0)
- [ ] Critical and brittle code paths are covered by unit tests
___

### Reviewer's Checklist
#### This section is for reviewers to fill out.

- [ ] Automated test coverage is satisfactory
- [ ] PR is fully functional
- [ ] PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- [ ] External dependency files were updated if necessary (`yarn` and `pip`)
- [ ] Documentation is updated
- [ ] Contributor is in AUTHORS.md

MisRob reviewed (changes requested):
I'm sorry @shivam-daksh but we don't want to replace `KExternalLink` by `ActionLink` as `ActionLink` is supposed to be gradually replaced by [Kolibri Design System ](https://design-system.learningequality.org/)components. That doesn't need to be done in scope of https://github.com/learningequality/studio/issues/4606, however we don't want move in the opposite direction.

I left a pretty detailed instructions in the "Guidance" section of https://github.com/learningequality/studio/issues/4606 that will help you to resolve it while keeping `KExternalLink`. Please try to follow it or ask questions here if there's anything that's not clear and I will make sure to clear any doubts.

MisRob commented on docker-compose.yml:
These changes will need to be reverted

MisRob commented:
@shivam-daksh In addition to the guidance in the issue, also see https://design-system.learningequality.org/kexternallink/. Apologies, I realize it'd be better to link it sooner.

shivam-daksh commented:
Hi @MisRob, as you said, I've changed the `ActionLink` back to `KExternalLink` adding a prop 'openInNewTab`. This has automatically added the redirect icon. Also, I've reverted the changes from `docker-compose.yml` which was done automatically by `Prettier` extension in my IDE.

MisRob commented:
Thanks @shivam-daksh! The latest changes are aligned with the guidance and look well overall to me. I currently have a queue of several larger PRs so I will loop in my colleagues for more detailed review and manual confirmation.

MisRob commented:
Thanks @shivam-daksh! The latest changes are aligned with the guidance and look well overall to me. I currently have a queue of several larger PRs so I will loop in my colleagues for more detailed review and manual confirmation. @ozer550 and @AllanOXDi would you please give it a look and test when you have a moment? Here's the issue https://github.com/learningequality/studio/issues/4606. Thank you.

AllanOXDi commented:
Hi @shivam-daksh Thanks for woking on this . links leading outside Studio now have a pop out icon. I  am leaving a few  clean up comments

AllanOXDi commented on contentcuration/contentcuration/frontend/settings/pages/UsingStudio/index.vue:
```diff
@@ -141,4 +145,7 @@
   }
-</style>
+  .kexternal-redirect{
+    margin-left: -8px;
```
The negative margin moves the element slightly to the left of its normal position. This could cause misalignment with other elements on the page, especially if the layout is responsive or if the link appears in different contexts.

shivam-daksh replied:
Hey @AllanOXDi , I think issue lies in `KExternalLink` component. The component by default is adding an inline 8px left margin. That's why I added -8px to realign back. [For More](https://github.com/learningequality/kolibri-design-system/blob/d3865313e65eab24195a2b65e411506b364ec9ba/lib/buttons-and-links/KExternalLink.vue#L106)

<img width="1395" alt="image" src="https://github.com/user-attachments/assets/13a4dfee-e3c6-4c3a-809c-f5edf8bf1106">

Unfortunately, I wasn't able to find other solutions for this.

MisRob replied:
Thanks both. @shivam-daksh I think it's fine to remove `margin-left: -8px;` from here, as @AllanOXDi suggests.

If we feel it's important, we can consider making adjustments directly to `KExternalLink` in the `kolibri-design-system` repository in the future, but for now I would just use the default behavior since it will be consistent with other links.

shivam-daksh replied:
Sure @MisRob, I've removed the `margin-left:-8px` . But it is causing little indendation issue in Settings->About Studio page. 

<img width="835" alt="image" src="https://github.com/user-attachments/assets/33636263-b739-42a7-bc56-860e55ae5f91">

AllanOXDi replied:
Hi @shivam-daksh 
 you could try this
```
.kexternal-redirect {
  margin-left: 0 ;
}
```

AllanOXDi replied:
Actually this won't work -sorry

shivam-daksh replied:
It's okay. Let me know if we have a fix for now.

shivam-daksh replied:
Only way I think of now is add the class in **About Studio** page only.

‘.kexternal-redirect { margin-left: 0 ; }‘

MisRob replied:
That sounds good @shivam-daksh. You're right that it needs to be aligned better on this page, thanks for elaborating.

shivam-daksh replied:
So, should I do it?

MisRob replied:
Yes

shivam-daksh replied:
That's done.

MisRob replied:
Thanks!

AllanOXDi commented on contentcuration/contentcuration/frontend/settings/pages/UsingStudio/index.vue:
```diff
@@ -64,16 +66,18 @@
       href="https://github.com/learningequality/studio/issues"
       :text="$tr('issuesPageLink')"
-      target="_blank"
+      openInNewTab="true"
```
The` openInNewTab="true"` is used correctly. However, in Vue, boolean props can be passed without a value. So it could be simplified to just` openInNewTab`.
<img width="350" alt="Screenshot 2024-08-06 at 22 37 05" src="https://github.com/user-attachments/assets/52fad4a4-12a7-4489-a906-f251d4cdfee2">

shivam-daksh replied:
Thank you for letting me know. I've changed it to simplified.

AllanOXDi commented on contentcuration/contentcuration/frontend/settings/pages/Storage/index.vue:
```diff
@@ -160,5 +160,8 @@
-</style>
+  .kexternal-redirect{
+    margin-left: -8px;
+    display: inline !important;
```
The` !important` property might not be necessary if there are no conflicting styles. It's generally best to avoid !important unless absolutely necessary. In fact, if you use the !important  property , it will override all previous styling rules for that specific property on that element :)

shivam-daksh replied:
And `display` property was indeed unnecessary, so I've removed it.

AllanOXDi reviewed (approved):
Hi @shivam-daksh, thanks for your hard work on this. The work looks good! Let's just make sure that the linting checks are passing, and then we can merge the changes. You can run `yarn run lint-frontend:format`

MisRob commented:
@shivam-daksh We figured out that what you're seeing may be just confusing output of the command. Running `yarn run lint-frontend:format` likely changed some of your local files. Do you have some uncommited changes after running it? If yes, please commit them and that should help.

shivam-daksh commented:
> @shivam-daksh We figured out that what you're seeing may be just confusing output of the command. Running `yarn run lint-frontend:format` likely changed some of your local files. Do you have some uncommited changes after running it? If yes, please commit them and that should help.

Hi @MisRob, Initially, my working tree was clean. But when I run that command a single change occured in a file that I commited. My working tree was clean and I ran the command `yarn run lint-frontend:format` and again same error occured but without changing any files. 
But I'm thinking that it is missing a file `/Volumes/Macintosh HD - Data/Space A/Open-Source/studio/kolibri/core/assets/src/core-app/apiSpec.js` which is not present. Actually, there is no such directory as **kolibri** in studio. It is actually a totally different repository. [Kolibri](https://github.com/learningequality/kolibri)
```
(.venv) shivam@Shivam-2 studio % git status
On branch fix-issue-#4606
nothing to commit, working tree clean
(.venv) shivam@Shivam-2 studio % yarn run lint-frontend:format
yarn run v1.22.22
$ yarn run lint-frontend --write
$ kolibri-tools lint --pattern 'contentcuration/contentcuration/frontend/**/*.{js,vue,scss,less,css}' --ignore '**/dist/**,**/node_modules/**,**/static/**' --write
Error: ENOENT: no such file or directory, open '/Volumes/Macintosh HD - Data/Space A/Open-Source/studio/kolibri/core/assets/src/core-app/apiSpec.js'
    at Object.openSync (node:fs:590:3)
    at Object.readFileSync (node:fs:458:35)
    at specModule (/Volumes/Macintosh HD - Data/Space A/Open-Source/studio/node_modules/kolibri-tools/lib/apiSpecExportTools.js:37:26)
    at Object.<anonymous> (/Volumes/Macintosh HD - Data/Space A/Open-Source/studio/node_modules/kolibri-tools/lib/apiSpecExportTools.js:103:13)
    at Module._compile (node:internal/modules/cjs/loader:1198:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1252:10)
    at Module.load (node:internal/modules/cjs/loader:1076:32)
    at Function.Module._load (node:internal/modules/cjs/loader:911:12)
    at Module.require (node:internal/modules/cjs/loader:1100:19)
    at require (node:internal/modules/cjs/helpers:119:18) {
  errno: -2,
  syscall: 'open',
  code: 'ENOENT',
  path: '/Volumes/Macintosh HD - Data/Space A/Open-Source/studio/kolibri/core/assets/src/core-app/apiSpec.js'
}
Kolibri Linter: Rewriting a prettier version of contentcuration/contentcuration/frontend/settings/pages/Storage/index.vue
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
(.venv) shivam@Shivam-2 studio % git status
On branch fix-issue-#4606
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   contentcuration/contentcuration/frontend/settings/pages/Storage/index.vue

no changes added to commit (use "git add" and/or "git commit -a")
(.venv) shivam@Shivam-2 studio % git add .
(.venv) shivam@Shivam-2 studio % git commit -m "yarn lint-frontend:format changes"          
[fix-issue-#4606 922b49e08] yarn lint-frontend:format changes
 1 file changed, 1 insertion(+)
(.venv) shivam@Shivam-2 studio % yarn run lint-frontend:format                    
yarn run v1.22.22
$ yarn run lint-frontend --write
$ kolibri-tools lint --pattern 'contentcuration/contentcuration/frontend/**/*.{js,vue,scss,less,css}' --ignore '**/dist/**,**/node_modules/**,**/static/**' --write
Error: ENOENT: no such file or directory, open '/Volumes/Macintosh HD - Data/Space A/Open-Source/studio/kolibri/core/assets/src/core-app/apiSpec.js'
    at Object.openSync (node:fs:590:3)
    at Object.readFileSync (node:fs:458:35)
    at specModule (/Volumes/Macintosh HD - Data/Space A/Open-Source/studio/node_modules/kolibri-tools/lib/apiSpecExportTools.js:37:26)
    at Object.<anonymous> (/Volumes/Macintosh HD - Data/Space A/Open-Source/studio/node_modules/kolibri-tools/lib/apiSpecExportTools.js:103:13)
    at Module._compile (node:internal/modules/cjs/loader:1198:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1252:10)
    at Module.load (node:internal/modules/cjs/loader:1076:32)
    at Function.Module._load (node:internal/modules/cjs/loader:911:12)
    at Module.require (node:internal/modules/cjs/loader:1100:19)
    at require (node:internal/modules/cjs/helpers:119:18) {
  errno: -2,
  syscall: 'open',
  code: 'ENOENT',
  path: '/Volumes/Macintosh HD - Data/Space A/Open-Source/studio/kolibri/core/assets/src/core-app/apiSpec.js'
}
✨  Done in 44.30s.
(.venv) shivam@Shivam-2 studio % git status
On branch fix-issue-#4606
nothing to commit, working tree clean

```

MisRob commented:
@shivam-daksh Thanks for following up. So the first step now - can you push the commited file just to see if that resolves the check on this PR?

MisRob commented:
@shivam-daksh Oh you already did, sorry I missed that! Re-running the checks now.

MisRob commented:
@shivam-daksh Linting passes now ;) So all good in the scope of this PR. Thanks for reporting the outputs, we took a note of it. Also thanks @AllanOXDi and @Pursottam6003, I think it's in good shape now thanks to your feedback. I skimmed briefly through the latest changes too and as soon as all checks pass, seems ready for merge to me.

MisRob commented:
Congrats to your first contribution @shivam-daksh and thanks for following-up!

AllanOXDi commented:
Well done :tada: @shivam-daksh

shivam-daksh commented:
Thank you, @MisRob, for your guidance throughout this process!
I appreciate your patience and support in helping me navigate my first contribution. I'm glad we were able to resolve the issues, and it feels great to see everything in good shape. Also, a big thanks to @AllanOXDi, @ozer550  and @Pursottam6003 for their valuable feedback.
Looking forward to contributing more and learning along the way. Thanks again for the encouragement and for making this a smooth experience!
